File: core/src/main/java/moe/caa/multilogin/core/ohc/RetryInterceptor.java
package moe.caa.multilogin.core.ohc;

import moe.caa.multilogin.api.internal.logger.LoggerProvider;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

/**
 * 延时重试拦截器
 */
public class RetryInterceptor implements Interceptor {
    private final int retry;
    private final long delay;

    public RetryInterceptor(int retry, long delay) {
        this.retry = retry;
        this.delay = delay;
    }

    @NotNull
    @Override
    public Response intercept(@NotNull Chain chain) throws IOException {
        Request request = chain.request();
        Response response;
        int tc = 0;
        while (true) {
            try {
                response = chain.proceed(request);
                return response;
            } catch (IOException e) {
                LoggerProvider.getLogger().debug(tc + " retry failed.", e);
                if (tc >= retry) throw e;
            }

            try {
                TimeUnit.MILLISECONDS.sleep(delay);
            } catch (InterruptedException e) {
                throw new InterruptedRetryException(e);
            }
            tc++;
            LoggerProvider.getLogger().debug("--> " + tc + " retry.");
        }
    }
}

File: core/src/main/kotlin/moe/caa/multilogin/core/intercept/RetryInterceptor.kt
package moe.caa.multilogin.core.intercept

import moe.caa.multilogin.core.util.logDebug
import okhttp3.Interceptor
import okhttp3.Response
import java.io.IOException

class RetryInterceptor(
    private val maxRetries: Int,
    private val retryDelay: Int
) : Interceptor {
    private var retries = 0

    override fun intercept(chain: Interceptor.Chain): Response {
        while (true) {
            try {
                return chain.proceed(chain.request())
            } catch (e: IOException) {
                if (maxRetries <= retries) throw e
                retries++
                logDebug("Request failed, retrying ($retries/$maxRetries)...", e)
                if (retryDelay > 0) Thread.sleep(retryDelay.toLong())
            }
        }
    }
}